## Introduction to Digital Earth Australia

#### [Digital Earth Australia](https://www.ga.gov.au/dea) (DEA) is a digital platform that offer large amounts of Earth observation data covering continental Australia and ready to be analysed.

#### DEA functionality is offered using a Python module, building on top and expanding the functionality that we have seen in Numpy, Matplotlib and XArray. To access the functionality of DEA we need to import the library. Then we declare a `Datacube` object that is going to be our point of access to the DEA data collections.

In [2]:
%matplotlib inline

import datacube

dc = datacube.Datacube(app='welcome-to-dea')

print(dc)
print("---------------")
type(dc)

Datacube<index=Index<db=PostgresDb<engine=Engine(postgresql://sandboxreader:***@db-aurora-dea-sandbox-eks.cluster-ro-cos5zfpkso9m.ap-southeast-2.rds.amazonaws.com:5432/sandbox)>>>
---------------


datacube.api.core.Datacube

#### Digital Earth Australia offers a catalogue of data from a range of satellite sensors that offer the most recent images all the way back to 1986. The DEA data comes from the following satellites:

* [Landsat 5 TM](https://www.usgs.gov/land-resources/nli/landsat/landsat-5?qt-science_support_page_related_con=0#qt-science_support_page_related_con) (LS5 TM), operational between March 1984 and January 2013
* [Landsat 7 ETM+](https://www.usgs.gov/land-resources/nli/landsat/landsat-7?qt-science_support_page_related_con=0#qt-science_support_page_related_con) (LS7 ETM+), operational since April 1999
* [Landsat 8 OLI](https://www.usgs.gov/land-resources/nli/landsat/landsat-8?qt-science_support_page_related_con=0#qt-science_support_page_related_con) (LS8 OLI), operational since February 2013
* [Sentinel 2A MSI](https://sentinel.esa.int/web/sentinel/missions/sentinel-2) (S2A MSI), operational since June 2015
* [Sentinel 2B MSI](https://sentinel.esa.int/web/sentinel/missions/sentinel-2) (S2B MSI, operational since March 2017

#### Landsat missions are jointly operated by the United States Geological Survey (USGS) and National Aeronautics and Space Administration (NASA). Sentinel missions are operated by the European Space Agency (ESA). One major difference between the two programs is the spatial resolution: each Landsat pixel represents 30 x 30 m on the ground while each Sentinel-2 pixel represents 10 x 10 m to 60 x 60 m depending on the spectral band.

![Image](https://prd-wret.s3-us-west-2.amazonaws.com/assets/palladium/production/s3fs-public/styles/full_width/public/thumbnails/image/dmidS2LS7Comparison.png)


#### The `Datacube` object that we have created has functionality attached to it. For example, we can get a list of the products available by doing:

In [5]:
dc.list_products().head()

name  \
id                                   
75  camden_insar_alos_displacement   
10     fc_percentile_albers_annual   
11   fc_percentile_albers_seasonal   
70                   ga_ls5t_ard_3   
71                   ga_ls7e_ard_3   

                                          description  gqa_stddev_xy  \
id                                                                     
75                     Camden InSAR ALOS Displacement            NaN   
10  Landsat Fractional Cover percentile 25 metre, ...            NaN   
11  Landsat Fractional Cover percentile 25 metre, ...            NaN   
70                  Landsat 5 TM ARD, GA Collection 3            NaN   
71                Landsat 7 ETM+ ARD, GA Collection 3            NaN   

    gqa_abs_iterative_mean_xy  gqa_cep90   instrument  gqa_mean_x  \
id                                                                  
75                        NaN        NaN         None         NaN   
10                        NaN        NaN  TM,ETM+,OLI         NaN   
11                        NaN        NaN  TM,ETM+,OLI         NaN   
70                        NaN        NaN         None         NaN   
71                        NaN        NaN         None         NaN   

    dataset_maturity  gqa_iterative_mean_xy  eo_gsd  ... gqa_abs_xy  gqa  \
id                                                   ...                   
75               NaN                    NaN     NaN  ...        NaN  NaN   
10               NaN                    NaN     NaN  ...        NaN  NaN   
11               NaN                    NaN     NaN  ...        NaN  NaN   
70               NaN                    NaN     NaN  ...        NaN  NaN   
71               NaN                    NaN     NaN  ...        NaN  NaN   

    eo_sun_azimuth  gqa_abs_x gqa_iterative_mean_x  creation_time        crs  \
id                                                                             
75             NaN        NaN                  NaN           None        NaN   
10             NaN        NaN                  NaN           None  EPSG:3577   
11             NaN        NaN                  NaN           None  EPSG:3577   
70             NaN        NaN                  NaN           None        NaN   
71             NaN        NaN                  NaN           None        NaN   

    resolution             tile_size  spatial_dimensions  
id                                                        
75         NaN                   NaN                 NaN  
10   (-25, 25)  (100000.0, 100000.0)              (y, x)  
11   (-25, 25)  (100000.0, 100000.0)              (y, x)  
70         NaN                   NaN                 NaN  
71         NaN                   NaN                 NaN  

[5 rows x 52 columns]

> Note: The `.head()` method in the previous cell limits the rown of the list to the first 10. You can try removing or commenting out this method using the `#` symbol to see the total number of rows although Jupyter will limit the size of the list. 

#### Exercise 5.1: EPSG stands for European Petroleum Survey Group and the accronym commonly refers to the geodetic parameter database with standard codes for coordinate systems, datums, spheroids and units. This code indicates the map projection. Can you find the EPSG code that DEA uses to store its data collections. Youn migth need to scroll towards the right side of the table to locate this column. 

* **a)** EPSG:4326
* **b)** EPSG:3577
* **c)** EPSG:2360

> Tip: If you are interested to know more about EPSG codes [here](https://spatialreference.org/) there is a a full list with descriptions.

#### Digital Earth Australia produces Analysis Ready Data (ARD) for each of the sensors listed above. The [ARD standard](http://ceos.org/ard/) for satellite data requires that data have undergone a number of processing steps, along with the creation of additional attributes for the data. DEA's ARD datasets include the following characteristics:

* **Geometric correction:** This includes establishing ground position, accounting for terrain (orthorectification) and ground control points, and assessing absolute position accuracy. 
Geometric calibration means that imagery is positioned accurately on the Earth's surface and stacked consistently so that sequential observations can be used to track meaningful change over time.
Adjustments for ground variability typically use a Digital Elevation Model (DEM).
* **Surface reflectance correction:** This includes adjustments for sensor/instrument gains, biases and offsets, include adjustments for terrain illumination and sensor viewing angle with respect to the pixel position on the surface.
Once satellite data is processed to surface reflectance, pixel values from the same sensor can be compared consistently both spatially and over time.
* **Observation attributes:** Per-pixel metadata such as quality flags and content attribution that enable users to make informed decisions about the suitability of the products for their use. For example, clouds, cloud shadows, missing data, saturation and water are common pixel level attributes.
* **Metadata:** Dataset metadata including the satellite, instrument, acquisition date and time, spatial boundaries, pixel locations, mode, processing details, spectral or frequency response and grid projection.


#### Optical sensors mounted on satellites measure the energy that comes from the sun and has been reflected by the Earth's surface. The sensor measures the energy in each of its spectral bands. There are many factors that affect the amount of energy that reaches the satellite sensor such as the solar and satellite angles relative to the surface and the atmospheric conditions. In order to analyse and compare images taken at different times, the data needs to be normalised into some standard conditions. Surface reflectance is one of this normalisations and is calculated using advanced physical models to correct the reflectance values taking into account atmospheric conditions, sun and satellite angles and local topography.

#### DEA offers to options for adjusted or normalised surface reflectance:

* **NBAR**: NBAR stands for *Nadir-corrected BRDF Adjusted Reflectance*, where BRDF stands for *Bidirectional reflectance distribution function*. The approach involves atmospheric correction to compute bottom-of-atmosphere radiance, and bi-directional reflectance modelling to remove the effects of topography and angular variation in reflectance. NBAR can be useful for analyses in extremely flat areas not affected by terrain shadow, and for producing attractive data visualisations that are not affected by NBART's nodata gaps (see below).

* **NBART**: NBART has the same features of NBAR but includes an additional *terrain illumination* reflectance correction that takes into account the surface topography. Terrain affects optical satellite images in a number of ways; for example, slopes facing the sun receive more sunlight and appear brighter compared to those facing away from the sun. This correction is performed with a Digital Surface Model (DSM). NBART is typically the default choice for performing analysis that are consistent through time. However, it can be introduce distortion in flat areas due to the noise in DSM data.

![Comparison between NBAR and NBART](data/nbar_nbart_animation.gif)

#### The animation above demonstrates how the NBART correction results in a significantly more two-dimensional looking image that is less affected by terrain illumination and shadow. Black pixels in the NBART image represent areas of deep terrain shadow that can't be corrected as they're determined not to be viewable by either the sun or the satellite. These are represented by -999 `nodata` values in the data.

> Note: Remember that remote sensing data is often stored as `uint16` to save space and `NaN` is only available for floating point types, that is why we need a value `-999` in this case to designate not-a-number data.

#### The *Observation Attributes (OA)* are a suite of measurements included in DEA's analysis ready datasets. They are an assessment of each image pixel to determine if it is an unobscured, unsaturated observation of the Earth's surface, along with whether the pixel is represented in each spectral band. The OA product allows users to exclude pixels that do not meet the quality criteria for their analysis. The capacity to automatically exclude such pixels is essential for analysing any change over time, since poor-quality pixels can significantly alter the percieved change over time. The most common use of OA is for cloud masking, where users can choose to remove images that have too much cloud, or ignore the clouds within each satellite image. A demonstration of how to use cloud masking can be found in the [masking data](../Frequently_used_code/Masking_data.ipynb) notebook.

#### The OA suite of measurements include the following observation pixel-based attributes:

* Null pixels
* Clear pixels
* Cloud pixels
* Cloud shadow pixels
* Snow pixels
* Water pixels
* Terrain shaded pixels
* Spectrally contiguous pixels (i.e. whether a pixel contains data in every spectral band)

#### Also included is a range of pixel-based attributes related to the satellite, solar and sensing geometries:

* Solar zenith
* Solar azimuth
* Satellite view
* Incident angle
* Exiting angle
* Azimuthal incident
* Azimuthal exiting
* Relative azimuth
* Timedelta


#### DEA uses XArray as its underlying data model and defines a simple querying interface to select satellite data which returns as `Dataset` objects.

In [ ]:
%matplotlib inline
import matplotlib.pyplot as plt

import datacube

dc = datacube.Datacube(app='welcome-to-dea')

type(dc)

#### This `dc` object is the handle to all DEA's functionality. For example to list the products available:

In [ ]:
dc.list_products().head()

#### Still, to select data we have to use numpy indexes. Wouldn't it be convenient to be able to select data from the images using the coordinates of the pixels instead of their relative positions?

#### This is exactly what XArray solves! Let's see how it works:

In [ ]:
query={'lat': (-35.25, -35.35),
         'lon': (149.05, 149.17),
         'time':('2010-01-01', '2019-01-15')}

ds = dc.load(product='ls8_nbart_geomedian_annual', **query)

ds

#### To explore XArray we have a file containing some reflectance data of Canberra that has been generated using the DEA library.

#### The object that we get `ds` is a XArray `Dataset`, which in some ways is very similar to the dictionary that we created before, but with lots of convenient functionality available.

In [ ]:
ds[['red', 'green', 'blue']].isel(time=0).to_array().plot.imshow(robust=True, figsize=(6,6))

#### Now that we have access to any location in Australia, select any place that you like in Australia and create a NRG→RGB False colour image, where N is the 'near-infrared' `nir` channel.

_Be careful with the extents in your request. Cubes with sizes 0.1 degree are easily managed by the Sandbox but you can quickly run out of memory if you increase this extent._